File: packages/shared-types/src/billing.ts
/**
 * Контракты биллинга (ТЗ §11, §12). Зеркалятся Pydantic-схемами в
 * apps/engine/app/schemas/billing.py. Изменение — синхронно + запись в DECISIONS.
 */
import type { Currency } from './index'

export type ModuleKey =
  | 'geoportal'
  | 'arm-analytics'
  | 'express-valuation'
  | 'report-generator'
  | 'interactive-report'
  | 'appraiser-calculator'

export type AccessType = 'quota' | 'period' | 'free'
export type ProviderName = 'mock' | 'monobank' | 'liqpay'
export type ProviderStatus =
  | 'created'
  | 'processing'
  | 'hold'
  | 'success'
  | 'failure'
  | 'reversed'
  | 'expired'
export type MappedStatus = 'paid' | 'pending' | 'failed' | 'canceled'
export type OrderStatus =
  | 'new'
  | 'awaiting_invoice'
  | 'pending'
  | 'paid'
  | 'failed'
  | 'fulfilled'
  | 'canceled'
export type EntitlementStatus = 'active' | 'exhausted' | 'expired' | 'canceled' | 'past_due'
export type PaymentMethod = 'card' | 'invoice'

// ── Платёжный провайдер ─────────────────────────────────────────
export interface CreateInvoiceInput {
  orderId: number | string
  amountMinor: number
  currency: Currency
  reference: string
  description?: string
  redirectUrl: string
  webhookUrl: string
}

export interface CreateInvoiceResult {
  providerInvoiceId: string
  payUrl: string
}

export interface WebhookEvent {
  invoiceId: string
  status: string
  amountMinor?: number
  currency?: string
  reference?: string
}

// ── Документы безнала (engine /api/reports/*) ───────────────────
export interface DocParty {
  name: string
  edrpou?: string
  ipn?: string
  address?: string
  iban?: string
  bankName?: string
  vat?: string
}

export interface DocItem {
  title: string
  qty: number
  priceMinor: number
}

export interface InvoiceDocRequest {
  orderNumber: string
  date: string
  seller: DocParty
  buyer: DocParty
  items: DocItem[]
  totalMinor: number
  currency: Currency
}

export interface ActDocRequest extends InvoiceDocRequest {
  paidDate?: string
}

// ── Представление доступа для кабинета ──────────────────────────
export interface EntitlementView {
  moduleKey: ModuleKey
  accessType: AccessType
  allowed: boolean
  quotaRemaining?: number
  periodEnd?: string | null
}

File: apps/engine/app/schemas/billing.py
"""Pydantic-схемы биллинг-документов (ТЗ §11, §12).

Зеркалят TS-контракты из packages/shared-types/src/billing.ts. Изменение
контракта — синхронно в обоих местах + запись в DECISIONS.md.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class DocParty(BaseModel):
    """Сторона документа (продавец/покупатель)."""

    name: str
    edrpou: str = ""
    ipn: str = ""
    address: str = ""
    iban: str = ""
    bank_name: str = Field("", alias="bankName")
    vat: str = ""

    model_config = {"populate_by_name": True}


class DocItem(BaseModel):
    title: str
    qty: int = 1
    price_minor: int = Field(alias="priceMinor")

    model_config = {"populate_by_name": True}


class InvoiceDocRequest(BaseModel):
    """Запрос на рахунок-фактуру (безнал, B2B)."""

    order_number: str = Field(alias="orderNumber")
    date: str
    seller: DocParty
    buyer: DocParty
    items: list[DocItem]
    total_minor: int = Field(alias="totalMinor")
    currency: str = "UAH"

    model_config = {"populate_by_name": True}


class ActDocRequest(InvoiceDocRequest):
    """Запрос на акт виконаних робіт (после оплаты)."""

    paid_date: str = Field("", alias="paidDate")

    model_config = {"populate_by_name": True}
